### DISCUSSION TITLE
Reading samples from stream

### DISCUSSION THREAD

**Discussion by uutzinger** [pschatzmann/arduino-audio-driver#17] [Q&A] [answered]
When I want to read a number of samples from a stream I convert the number of samples to the corresponding number of bytes. Then I request them from the stream.

If the stream does not have the number of bytes available it will return the available bytes and let me know how many were available.
After a short delay, I will start another read request and ask for the missing number of bytes. I repeat this until I received all bytes I want.  

My questions are: 
- Is it possible that the stream returns a partial sample? E.g. With 16 bits per sample, is it possible I receive an odd number of bytes?
- Is it possible that the read function returns more bytes than I requested? If that could be the case I would need to leave extra space in my receiving buffer.

**Comment by pschatzmann** [comment, marked as answer]
In general: The API operates with bytes and not samples! 

When reading data it is recommended to use a multiple of the sample size * channels to make sure that full frames are received. In any case the result of the readBytes() should be used to determine the number of effectively returned bytes. But in most cases this corresponds to the requested number of bytes since I am using blocking reads in most of the cases. 

Since the bits per sample can be 16 or 32, good values would be 1024, 512, 256, 128 and 64 bytes since this will return full samples both for mono and stereo. The same applies when you have the possibility to define an buffer size. 

So to answer your question: I expect that you would only get an odd number of bytes when you explicitly requested it (which should be avoided) and it will never provide more data then requested.

Do you have a specific class in mind or did you see any other behavior ?

ps. This project is only here to configure the DAC (or ADC) and no audio data is involved: this is the responsibility of the I2S API.

**Reply by uutzinger** [reply]
Thanks very much for your answer.

I wanted to test ES8388 with custom serial print and process.

I decided that for my custom process it is more useful to contribute a function to audio-tools and I posted a proposed extension of BaseConverter.h in a separate message in the audio-tools discussion.
